Draft CI requirements (#61)
bemanproject/beman: docs/BEMAN_STANDARD.md

**[CPP.NAMESPACE]** RECOMMENDATION: Headers in `include/beman/<short_name>/` should export
entities in the `beman::<short_name>` namespace.

> ## Continous Integration

Reviewer: ```suggestion
## Continuous integration
```

> **[CI.PLATFORM]** REQUIREMENT: All CI should be implemented using GitHub Actions or an CI platform that appripriatly interacts with the code control system.

Author: `an CI platform that appropriately interacts with the code control system` may need better wording here.

Reviewer 2: If someone wants to pay for an external CI system, that might be acceptable? We can cross that bridge if and when. 

But for beman standards purposes, configuration as code is important, and that code should live with the project. That's built in with GH Actions, but it does mean I can mostly coax Gitea to use them, too. I haven't tried Gitlab. 

Internally, we use GHE, but not Actions, and instead use our own systems for CI and CD, but they're sufficiently agnostic that it's not an issue. It is why I prefer as minimal magic in the actions as possible, though, so it's straightforward to run tests on change.

Author: Yeah I agree, we need to come up with a wording for allowing acceptable external CI or make this item a recommendation.

> **[CI.PLATFORM]** REQUIREMENT: All CI should be implemented using GitHub Actions or an CI platform that appripriatly interacts with the code control system.

Reviewer: ```suggestion
**[CI.PLATFORM]** REQUIREMENT: The CI pipeline must be implemented using GitHub Actions.
```

- check REQUIREMENT and "must" in this docs
- I don't think we need to have external cases for start
- We should pick between `REQUIREMENT: The CI pipeline must be implemented using GitHub Actions.` and `RECOMMANDATION: The CI pipeline should be implemented using GitHub Actions. These options are consistent with current doc wording.

````

> **[CI.TEST]** REQUIREMENT: CI must execute appripriate testing of reasonable scope tagged with each commit, and must fail with test cases.

Reviewer 3: ```suggestion
**[CI.TEST]** REQUIREMENT: CI must execute appropriate testing of reasonable scope tagged with each commit, and must fail with test cases.
```
spelling....

Author: ops, thanks.

> **[CI.TEST]** REQUIREMENT: CI must execute appropriate testing of reasonable scope tagged with each commit, and must fail with test cases.

Reviewer: ```suggestion
**[CI.ENABLE_TESTS]** REQUIREMENT: The CI must execute tests for each commit in each PR and must provide explicit output on failure.
```

and I would remove

> **[CI.TEST.CROSS_PLATFORM]** REQUIREMENT: CI enabled tests must be executed across major platforms if appripriate.

Reviewer: We don't have multiple dot usage in the name of the rules in this doc!

+ remove `CI.TEST.CROSS_COMPILER_FAMILIES` and `CI.TEST.CROSS_COMPILER_VERSIONS`

```suggestion
**[CI.USE_CROSS_PLATFORM_TESTS]** RECOMMANDATION: The CI tests should be executed across major platforms and compilers.
```

**[CI.TEST.CROSS_COMPILER_FAMILIES]** REQUIREMENT: CI enabled tests must be executed across major compiler families if appripriate.

> 

Reviewer 3: These are listed as Requirements, but then come with the conditional 'if appropriate' -- which kinda makes them recommendations I'd say.

Author: This is fair.

My thinking here is, let's say we have a feature that relies on trunk version of gcc/ clang, this requirement could be waived.

**[CI.TEST.CROSS_COMPILER_VERSIONS]** RECOMMENDATION: CI enabled tests should be executed across support major versions of appripriate compiler families.

> **[CI.DURATION_PER_COMMIT]** RECOMMENDATION: CI pipeline associated with commits should finish in a reasonable amount of time.

Author: Should we include a quantifiable "reasonable amount of time" as suggestion here?

Author: My recommendation is 10 mins for small projects, no longer than 30 mins.

> **[CI.DURATION_PER_COMMIT]** RECOMMENDATION: CI pipeline associated with commits should finish in a reasonable amount of time.

Reviewer: ```suggestion
**[CI.EXECUTION_TIME]** RECOMMENDATION: The entire CI pipeline should run as fast as possible.
```

Again, I would make it short or remote it.

> **[CI.SCHEDULED_TEST]** RECOMMENDATION: CI enabled checks should run periodically across the entire repository.

Reviewer: I would not put this rule in the current PR, as we don't have yet implemented in any repo. Please remove for now.

> **[CI.SCEDULED_TEST_NOTIFY_FAILURE]** REQUIREMENT: There must be a established protocol to report CI failure (e.g. opening a GitHub issue) when scheduled CI fails.

Author: Alternative wording: A CI run must not fail silently.

Reviewer 3: I probably don't want an issue opened every time there's a failure...

Reviewer 2: While it's very useful to know the state of success across compilers, I think it would be fine for say, AppleClang, to not be able to compile something because it's too old to have some feature that's needed, or even just desired, for the library component.

Requiring C++26 is OK. In fact it may be necessary for adoption of some proposals.

Reviewer: I would not put this rule in the current PR, as we don't have yet implemented in any repo. Please remove for now.
